### PR TITLE
ARROW-49: [Python] Add Column and Table wrapper interface

### PR DESCRIPTION
**Pull request by xhochy**
After https://github.com/apache/arrow/pull/52 is merged, I'd like to split Column and Table into separate .pyx files, array.pyx seems a bit overcrowded.

### CODE REVIEW

**Comment by wesm** [comment]
Yep — that can all move to a `table.pyx` file if that works for you

File: python/pyarrow/array.pyx [hunk 1]
```diff
@@ -360,3 +425,35 @@ cdef class Table:
...
         return pd.DataFrame(dict(zip(names, data)), columns=names)
+
+    def name(self):
+        self._check_nullptr()
+        return frombytes(self.table.name())
+
+    def schema(self):
+        raise NotImplementedError()
+
+    def column(self, index_or_name):
```
Comment (wesm) [inline python/pyarrow/array.pyx]: I would prefer only to pass indexes here -- retrieval by name can go in a different method. Can you also add a `__getitem__(self, i)` that invokes `self.column(i)`?

File: python/pyarrow/array.pyx [hunk 2]
```diff
@@ -360,3 +425,35 @@ cdef class Table:
             data.append(<object> arr)
 
         return pd.DataFrame(dict(zip(names, data)), columns=names)
+
+    def name(self):
```
Comment (wesm) [inline python/pyarrow/array.pyx]: Make this `property name:`

File: python/pyarrow/array.pyx [hunk 3]
```diff
@@ -287,6 +287,66 @@ cdef class RowBatch:
...
+
+        def __get__(self):
+            self._check_nullptr()
+            return (self.length(),)
+
+    def null_count(self):
+        self._check_nullptr()
+        return self.column.null_count()
+
+    def name(self):
```
Comment (wesm) [inline python/pyarrow/array.pyx]: Property

File: python/pyarrow/array.pyx [hunk 4]
```diff
@@ -287,6 +287,66 @@ cdef class RowBatch:
...
+        self._check_nullptr()
+        return self.column.length()
+
+    property shape:
+
+        def __get__(self):
+            self._check_nullptr()
+            return (self.length(),)
+
+    def null_count(self):
```
Comment (wesm) [inline python/pyarrow/array.pyx]: property

File: python/pyarrow/array.pyx [hunk 5]
```diff
@@ -287,6 +287,66 @@ cdef class RowBatch:
...
+        self._check_nullptr()
+        return self.column.null_count()
+
+    def name(self):
+        return frombytes(self.column.name())
+
+    def type(self):
+        raise NotImplementedError("Type information not yet implemented")
+
+    def data(self):
```
Comment (wesm) [inline python/pyarrow/array.pyx]: These should probably also be properties

File: python/pyarrow/array.pyx [hunk 6]
```diff
@@ -287,6 +287,66 @@ cdef class RowBatch:
...
+    def _check_nullptr(self):
+        if self.column == NULL:
+            raise ReferenceError("Column object references a NULL pointer."
+                    "Not initialized.")
+
+    def __len__(self):
+        self._check_nullptr()
+        return self.column.length()
+
+    def length(self):
```
Comment (wesm) [inline python/pyarrow/array.pyx]: Property maybe? Not as important since we have `__len__`

File: python/pyarrow/array.pyx [hunk 7]
```diff
@@ -287,6 +287,66 @@ cdef class RowBatch:
...
+        """
+        cdef:
+            PyObject* arr
+
+        import pandas as pd
+
+        check_status(pyarrow.ArrowToPandas(self.sp_column, &arr))
+        return pd.Series(<object>arr, name=self.name())
+
+    def _check_nullptr(self):
```
Comment (wesm) [inline python/pyarrow/array.pyx]: This one can be cdef'd, unless we expect to call it from Python

File: python/pyarrow/array.pyx [hunk 8]
```diff
@@ -360,3 +425,35 @@ cdef class Table:
...
+
+    def columns(self):
+        for i in range(self.num_columns()):
+            yield self.column(i)
+
+    def num_columns(self):
+        self._check_nullptr()
+        return self.table.num_columns()
+
+    def num_rows(self):
```
Comment (wesm) [inline python/pyarrow/array.pyx]: Perhaps make these properties also -- can you add `__len__` that calls `num_rows`?

File: python/pyarrow/array.pyx [hunk 9]
```diff
@@ -360,3 +425,35 @@ cdef class Table:
...
+        self._check_nullptr()
+        index = index_or_name
+        # TODO(uwe): Implement retrevial by name
+        cdef Column column = Column()
+        column.init(self.table.column(index))
+        return column
+
+    def columns(self):
+        for i in range(self.num_columns()):
+            yield self.column(i)
```
Comment (wesm) [inline python/pyarrow/array.pyx]: If this returns an iterator, should call the method `itercolumns`.

**Comment by wesm** [comment]
Thank you, this is a good start. If you like, feel free to go ahead and move this to `pyarrow.table`

**Comment by xhochy** [comment]
Incorporated the comments and also added the missing ChunkedArray wrapper class. Now we should have wrapper for all Arrow classes used in the basic DataFrame conversions.

**Comment by wesm** [comment]
+1, thank you, this is excellent

**Pull request closed without merging**